### PR TITLE
Switching C++ to C MPI bindings for compatibility with OpenMPI > v2.0

### PR DESCRIPTION
**Pull request by alistaireverett**
Changes to address issue #175. 

This appears to compile and run fine for me, but could do with a sanity check (and there may be more elegant ways of implementing some of the changes!).

In most cases it's just a straight swap for C++ to C MPI bindings, in a few cases things had to be shuffled around to make sure variables were declared/assigned correctly - these are probably the bits where there might be some room for improvement.

### CODE REVIEW

File: horizontal_adaptivity/tests/test_main.cpp [hunk 1]
```diff
@@ -51,18 +51,19 @@ int main(int argc, char **argv) {
...
 #ifdef HAVE_MPI
-  MPI::Init(argc, argv);
+  MPI_Init(&argc, &argv);
   // Undo some MPI init shenanigans
   chdir(getenv("PWD"));
 #endif
 #ifdef HAVE_PETSC  
   PetscInitialize(&argc, &argv, NULL, PETSC_NULL);
   // PetscInitializeFortran needs to be called when initialising PETSc from C, but calling it from Fortran
+  // This sets all kinds of objects such as PETSC_NULL_OBJECT, PETSC_COMM_WORLD, etc., etc.
```
Comment (jrper) [inline horizontal_adaptivity/tests/test_main.cpp]: Additional comment reinserted.

File: femtools/tests/test_main.cpp [hunk 2]
```diff
@@ -57,13 +57,14 @@ int main(int argc, char **argv)
...
 #ifdef HAVE_MPI
-  MPI::Init(argc, argv);
+  MPI_Init(&argc, &argv);
   // Undo some MPI init shenanigans
   chdir(getenv("PWD"));
 #endif
 #ifdef HAVE_PETSC  
   PetscInitialize(&argc, &argv, NULL, PETSC_NULL);
   // PetscInitializeFortran needs to be called when initialising PETSc from C, but calling it from Fortran
+  // This sets all kinds of objects such as PETSC_NULL_OBJECT, PETSC_COMM_WORLD, etc., etc.
```
Comment (jrper) [inline femtools/tests/test_main.cpp]: Additional comment reinserted.

File: libadaptivity/load_balance/include/comTools.h [hunk 3]
```diff
@@ -215,18 +217,19 @@ void allSendRecv(const std::vector< std::vector<T> >& sendBuffer,
...
-  MPI::Status recvStatus;
+  std::vector<MPI_Request> sendRequest(NProcs, MPI_REQUEST_NULL);
+  std::vector<MPI_Request> recvRequest(NProcs, MPI_REQUEST_NULL);
+  MPI_Status recvStatus;
   
   // Send
   for(size_t p=0;p<NProcs;p++){
     if(send_count[p]){
-      sendRequest[p]=MPI::COMM_WORLD.Isend(&(sendBuffer[p][0]), send_count[p], const_mpi_type, p, 1);
+      //sendRequest[p]=MPI_COMM_WORLD.Isend(&(sendBuffer[p][0]), send_count[p], const_mpi_type, p, 1);
```
Comment (jrper) [inline libadaptivity/load_balance/include/comTools.h]: remove C++ call left as comment.

File: libadaptivity/load_balance/include/comTools.h [hunk 4]
```diff
@@ -215,18 +217,19 @@ void allSendRecv(const std::vector< std::vector<T> >& sendBuffer,
...
+      MPI_Isend(&(sendBuffer[p][0]), send_count[p], const_mpi_type, p, 1,MPI_COMM_WORLD, &sendRequest[p]);
```
Comment (jrper) [inline libadaptivity/load_balance/include/comTools.h]: I see they are't needed, but brackets around the new reference might make it less ambiguous.

File: libadaptivity/load_balance/include/comTools.h [hunk 5]
```diff
@@ -237,13 +240,12 @@ void allSendRecv(const std::vector< std::vector<T> >& sendBuffer,
       recvBuffer[p].resize(recv_count[p]);
       
       // initiate non-blocking receive
-      recvRequest[p]=MPI::COMM_WORLD.Irecv(&(recvBuffer[p][0]), recv_count[p], const_mpi_type, p, 1);
+      MPI_Irecv(&(recvBuffer[p][0]), recv_count[p], const_mpi_type, p, 1, MPI_COMM_WORLD, &recvRequest[p]);
```
Comment (jrper) [inline libadaptivity/load_balance/include/comTools.h]: ditto.

File: libadaptivity/load_balance/src/Mesh.cpp [hunk 6]
```diff
@@ -423,32 +428,32 @@ void Mesh::halo_update(const vector< vector<unsigned> >& data_in, vector< vector
...
   for(int i=0;i<NProcs;i++){
     int cnt = stride*num_nodes_halo(i);
 
     if((i==MyRank)||(cnt==0)){
-      RecvRequest[i] =  MPI::REQUEST_NULL;
+      RecvRequest[i] =  MPI_REQUEST_NULL;
     }else{    
       data_out[i].resize( cnt );
-      RecvRequest[i] = MPI::COMM_WORLD.Irecv(&(data_out[i][0]), cnt, MPI::UNSIGNED, i, 13);
+      MPI_Irecv(&(data_out[i][0]), cnt, MPI_UNSIGNED, i, 13, MPI_COMM_WORLD, &RecvRequest[i]);
```
Comment (jrper) [inline libadaptivity/load_balance/src/Mesh.cpp]: Brackets round RecvRequest[i].

File: libadaptivity/load_balance/src/Mesh.cpp [hunk 7]
```diff
@@ -423,32 +428,32 @@ void Mesh::halo_update(const vector< vector<unsigned> >& data_in, vector< vector
...
+  vector<MPI_Request> SendRequest(NProcs);
   for(int i = 0; i<NProcs; i++){
     int cnt = stride*num_nodes_shared(i);
 
     if((i==MyRank)||(cnt==0)){
-      SendRequest[i] =  MPI::REQUEST_NULL;
+      SendRequest[i] =  MPI_REQUEST_NULL;
     }else{    
-      SendRequest[i] = MPI::COMM_WORLD.Isend(&(data_in[i][0]), cnt, MPI::UNSIGNED, i, 13);
+      MPI_Isend(&(data_in[i][0]), cnt, MPI_UNSIGNED, i, 13, MPI_COMM_WORLD, &SendRequest[i]);
```
Comment (jrper) [inline libadaptivity/load_balance/src/Mesh.cpp]: Brackets.

File: libadaptivity/load_balance/src/Mesh.cpp [hunk 8]
```diff
@@ -470,33 +475,33 @@ void Mesh::halo_update(const vector< vector<unsigned> >& data_in, const vector<u
...
   for(int i = 0; i<NProcs; i++){
     int cnt = stride*data_cnt[i];
     
     if((i==MyRank)||(cnt==0)){
-      RecvRequest[i] =  MPI::REQUEST_NULL;
+      RecvRequest[i] =  MPI_REQUEST_NULL;
     }else{
       data_out[i].resize(cnt);
-      RecvRequest[i] = MPI::COMM_WORLD.Irecv(&(data_out[i][0]), cnt, MPI::UNSIGNED, i, 13);
+      MPI_Irecv(&(data_out[i][0]), cnt, MPI_UNSIGNED, i, 13, MPI_COMM_WORLD, &RecvRequest[i]);
```
Comment (jrper) [inline libadaptivity/load_balance/src/Mesh.cpp]: Brackets.

File: libadaptivity/load_balance/src/Mesh.cpp [hunk 9]
```diff
@@ -516,33 +521,33 @@ void Mesh::halo_update(const vector< vector<samfloat_t> >& data_in, vector< vect
...
   for(int i = 0; i<NProcs; i++){
     int cnt = stride*num_nodes_halo(i);
     
     if((cnt==0)||(i==MyRank)){
-      RecvRequest[i] =  MPI::REQUEST_NULL;
+      RecvRequest[i] =  MPI_REQUEST_NULL;
     }else{    
       data_out[i].resize( cnt );
-      RecvRequest[i] = MPI::COMM_WORLD.Irecv(&(data_out[i][0]), cnt, MPI::SAMFLOAT, i, 13);
+      MPI_Irecv(&(data_out[i][0]), cnt, SAMFLOAT, i, 13, MPI_COMM_WORLD, &RecvRequest[i]);
```
Comment (jrper) [inline libadaptivity/load_balance/src/Mesh.cpp]: Brackets

File: libvtkfortran/vtkfortran.cpp [hunk 10]
```diff
@@ -738,18 +734,8 @@ extern "C" {
...
-    if (!writer->GetController()) {
-      vtkMPIController *cont = vtkMPIController::New();
-      cont->SetCommunicator(vtkMPICommunicator::GetWorldCommunicator());
-      writer->SetController(cont);
-    }
-    writer->SetWriteSummaryFile(true);
-#else
     writer->SetWriteSummaryFile((*rank)==0);
-#endif
+
```
Comment (jrper) [inline libvtkfortran/vtkfortran.cpp]: New code missing (merge problem?)

File: tools/petsc_readnsolve_main.cpp [hunk 11]
```diff
@@ -123,6 +125,7 @@ int main(int argc, char **argv){
   static char help[] = "Use -help to see the help.\n\n";
   PetscErrorCode ierr = PetscInitialize(&argc, &argv, NULL, help);
   // PetscInitializeFortran needs to be called when initialising PETSc from C, but calling it from Fortran
+  // This sets all kinds of objects such as PETSC_NULL_OBJECT, PETSC_COMM_WORLD, etc., etc.
```
Comment (jrper) [inline tools/petsc_readnsolve_main.cpp]: Old comment reinserted.

File: tools/test_pressure_solve_main.cpp [hunk 12]
```diff
@@ -43,6 +43,7 @@ int main(int argc, char **argv){
   static char help[] = "Use -help to see the help.\n\n";
   PetscErrorCode ierr = PetscInitialize(&argc, &argv, NULL, help);
   // PetscInitializeFortran needs to be called when initialising PETSc from C, but calling it from Fortran
+  // This sets all kinds of objects such as PETSC_NULL_OBJECT, PETSC_COMM_WORLD, etc., etc.
```
Comment (jrper) [inline tools/test_pressure_solve_main.cpp]: Old comment reinserted.

**Comment by tmbgreaves** [comment]
Very many thanks for this, Alistair! With our Bionic builder currently breaking gfortran I guess putting up a Fedora builder is a reasonable test for this, and independently useful for early warnings on centos. I'll aim to do that this week along with the centos6 build fixes.

**Comment by alistaireverett** [comment]
Thanks for the comments, hopefully the new commits fix all of those. Also pulled in the latest changes from the trunk and corrected the old merge issues which crept in.

**Comment by jrper** [comment]
Great, it looks fine to me now, so we're just waiting on buildbot.

**Comment by tmbgreaves** [comment]
Working on a Fedora 28 environment now.

**Comment by tmbgreaves** [comment]
Referencing James' recent comments in issue #192 Fedora 28 might not be a good option given he's seen build issues with gcc 8 there but not with Fedora 27. I'll go with 27 here.

**Comment by tmbgreaves** [comment]
Thinking this through - we have no guarantee that exotic (read - Fedora) builds will work irrespective of this merge, so a better test is probably 'does it all work with 16.04 LTS' for acceptance at this point. Queue set up accordingly, and will keep an eye on it. If no objections from others, I'll merge this if/when xenial passes.

**Comment by jrper** [comment]
Fine by me.

**Comment by tmbgreaves** [comment]
Just the one unit test unhappy as below. Anyone with more petsc knowledge than me - is this likely to be something fixable in Fluidity or something that needs a petsc rebuild with changed config?

    test_sam_integration: Running
SPUD WARNING: Creating __value child for non null element - deleting parent data
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see http://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[0]PETSC ERROR: to get more information on the crash.
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Signal received
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.6.3, Dec, 03, 2015 
[0]PETSC ERROR: ./test_sam_integration on a linux-gnu-c-opt named e6ae7bed33ea by Unknown Thu May 10 13:54:59 2018
[0]PETSC ERROR: Configure options --with-shared-libraries --with-debugging=0 --useThreads 0 --with-clanguage=C++ --with-c-support --with-fortran-interfaces=1 --with-mpi-dir=/usr/lib/openmpi --with-lapack-lib=-llapack --with-blas-lib=-lblas --with-blacs=1 --with-blacs-include=/usr/include --with-blacs-lib="[/usr/lib/libblacs-openmpi.so,/usr/lib/libblacs-openmpi.so]" --download-hypre=externals/hypre-2.10.0b-p2.tar.gz --download-ml=externals/ml-6.2-p3.tar.gz --with-fortran-interfaces=1 --download-ctetgen=externals/ctetgen-0.4.tar.gz --download-chaco=externals/Chaco-2.2-p2.tar.gz --download-metis=externals/metis-5.1.0-p1.tar.gz --download-mumps=externals/MUMPS_5.0.0-p1.tar.gz --download-scalapack=externals/scalapack-2.0.2.tgz --download-parmetis=externals/parmetis-4.0.3-p2.tar.gz --download-exodusii=externals/exodus-5.24.tar.bz2 --download-sowing=externals/sowing-1.1.17-p2.tar.gz --download-triangle=externals/Triangle.tar.gz --download-ptscotch=externals/scotch_6.0.3.tar.gz --download-suitesparse=externals/SuiteSparse-4.4.3.tar.gz --with-netcdf=1 --with-netcdf-dir=/usr --with-hdf5=1 --with-hdf5-include=/usr/include/hdf5/openmpi --with-hdf5-lib="[-lhdf5_openmpi,-lhdf5_openmpi_hl]" --CXX_LINKER_FLAGS=-Wl,--no-as-needed
[0]PETSC ERROR: #1 User provided function() line 0 in  unknown file
    test_sam_integration: --------------------------------------------------------------------------
    test_sam_integration: MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
    test_sam_integration: with errorcode 59.
    test_sam_integration: NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
    test_sam_integration: You may or may not see output from other processes, depending on
    test_sam_integration: exactly when Open MPI kills them.
    test_sam_integration: --------------------------------------------------------------------------
    test_sam_integration: PARMETIS ERROR: The sum of tpwgts for constraint #0 is not 1.0
    WARNING: no output from test
    ERROR: non-zero exit code from test

**Comment by tmbgreaves** [comment]
Docker image pushed up to the hub if anyone needs it for local testing:  fluidity/buildbot:c-mpi-bindings-latest-pretest

**Comment by stephankramer** [comment]
That unittest is probably the only thing that currently tests the sam load balancer (the code that lives in libadaptivity/load_balance/), which has been our sort of fallback for when zoltan isn't available. @tmbgreaves : Do we actually still currently test a build without zoltan that *does* run the parallel tests (but not the parallel test with the zoltan tag)? If we don't it's very likely that sam is broken generally, so we could consider purging it...

**Comment by tmbgreaves** [comment]
No, we dropped the sam build a while back. Purging sam sounds like something well outside this PR's remit - any objections if I drop that unittest for the purposes of merging?

**Comment by stephankramer** [comment]
That sounds like a good solution to me then, yes

**Comment by jrper** [comment]
Running valgrind suggests that there is a bug being introduced to SAM, which appears to be slightly subtle and due to shadowing of class variables in `libadaptivity/load_balance/src/MI5.cpp` and  `libadaptivity/load_balance/src/packing.cpp`. @tmbgreaves if I fix this, do you want me to turn the test  back on or not?

**Comment by tmbgreaves** [comment]
@jrper - sounds like there is no harm in turning it back on. Just prune out a510bc and force-push with your fix sounds good.

**Comment by tmbgreaves** [comment]
I've started a rerun

**Comment by tmbgreaves** [comment]
Looking good through unittests; I'll let the run complete overnight and merge assuming all goes green.

**Comment by alistaireverett** [comment]
Thanks for catching those @jrper

**Pull request merged**